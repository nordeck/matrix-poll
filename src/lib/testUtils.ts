/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  PowerLevelsStateEvent,
  RoomEvent,
  RoomMemberStateEventContent,
  StateEvent,
} from '@matrix-widget-toolkit/api';
import {
  GroupContent,
  IPoll,
  IPollSettings,
  IVote,
  PollStartEvent,
  PollType,
  ResultType,
  RoomNameEvent,
} from '../model';

/**
 * Create a matrix room member event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockRoomMember({
  state_key = '@user-alice',
  event_id = '$event-id-0',
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  content?: Partial<RoomMemberStateEventContent>;
} = {}): StateEvent<RoomMemberStateEventContent> {
  return {
    type: 'm.room.member',
    sender: '@user-id',
    content: {
      membership: 'join',
      displayname: 'Alice',
      avatar_url: 'mxc://alice.png',
      ...content,
    },
    state_key,
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a matrix room name event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockRoomName({
  room_id = '!room-id',
  content = {},
}: {
  room_id?: string;
  content?: Partial<RoomNameEvent>;
} = {}): StateEvent<RoomNameEvent> {
  return {
    type: 'm.room.name',
    sender: '',
    content: {
      name: 'My Room',
      ...content,
    },
    state_key: room_id,
    origin_server_ts: 0,
    event_id: '$event-id-0',
    room_id,
  };
}

/**
 * Create a matrix power levels event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPowerLevelsEvent({
  content = {},
}: {
  content?: Partial<PowerLevelsStateEvent>;
} = {}): StateEvent<PowerLevelsStateEvent> {
  return {
    type: 'm.room.power_levels',
    sender: '@user-id',
    content: {
      users_default: 100,
      ...content,
    },
    state_key: '',
    origin_server_ts: 0,
    event_id: '$event-id-0',
    room_id: '!room-id',
  };
}

/**
 * Create a poll event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPoll({
  state_key = 'poll-0',
  event_id = '$event-id-0',
  origin_server_ts = 0,
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  origin_server_ts?: number;
  content?: Partial<IPoll>;
} = {}): StateEvent<IPoll> {
  return {
    type: 'net.nordeck.poll',
    sender: '@user-id',
    content: {
      title: 'My Title',
      question: 'My Question',
      description: 'My Description',
      pollType: PollType.Open,
      answers: [
        { id: '1', label: 'Yes' },
        { id: '2', label: 'No' },
      ],
      resultType: ResultType.Visible,
      duration: 1,
      groups: [
        {
          id: 'red-party',
          eventId: '$event-id-0',
          abbreviation: 'Red Party',
          color: '#ff0000',
          votingRights: {
            '@user-alice': {
              state: 'active',
            },
            '@user-bob': {
              state: 'active',
            },
            '@user-charlie': {
              state: 'active',
            },
          },
        },
        {
          id: 'blue-party',
          eventId: '$event-id-1',
          abbreviation: 'Blue Party',
          color: '#0000ff',
          votingRights: {
            '@user-dameon': {
              state: 'active',
            },
          },
        },
      ],
      ...content,
    },
    state_key,
    origin_server_ts,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a poll start event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPollStart({
  event_id = '$start-event-id',
  origin_server_ts = 0,
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  origin_server_ts?: number;
  content?: Partial<PollStartEvent>;
} = {}): RoomEvent<PollStartEvent> {
  return {
    type: 'net.nordeck.poll.start',
    sender: '@user-id',
    content: {
      ...content,
    },
    origin_server_ts,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a group event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockGroup({
  state_key = 'group-0',
  event_id = '$event-id-0',
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  content?: Partial<GroupContent>;
} = {}): StateEvent<GroupContent> {
  return {
    type: 'net.nordeck.poll.group',
    sender: '@user-id',
    content: {
      abbreviation: 'GROUP 0',
      color: '#07f556',
      members: {},
      ...content,
    },
    state_key,
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a matrix vote event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockVote({
  event_id,
  sender = '@user-bob',
  origin_server_ts = 0,
  content = {},
}: {
  event_id?: string;
  sender?: string;
  origin_server_ts?: number;
  content?: Partial<IVote>;
} = {}): RoomEvent<IVote> {
  return {
    type: 'net.nordeck.poll.vote',
    sender,
    content: {
      answerId: '1',
      pollId: 'poll-0',
      'm.relates_to': {
        rel_type: 'm.reference',
        event_id: '$start-event-id',
      },
      ...content,
    },
    origin_server_ts,
    event_id: event_id ?? sender,
    room_id: '!room-id',
  };
}

/**
 * Create a poll settings event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPollSettings({
  event_id = '$event-id-0',
  content = {},
}: {
  event_id?: string;
  content?: Partial<IPollSettings>;
} = {}): StateEvent<IPollSettings> {
  return {
    type: 'net.nordeck.poll.settings',
    sender: '@user-id',
    content: {
      ...content,
    },
    state_key: '!room-id',
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}
