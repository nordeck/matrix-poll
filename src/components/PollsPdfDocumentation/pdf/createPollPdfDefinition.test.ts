/*
 * Copyright 2023 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mockPoll, mockRoomMember } from '../../../lib/testUtils';
import { createPollPdfDefinition } from './createPollPdfDefinition';

describe('createPollPdfDefinition', () => {
  const getUserDisplayName = (id: string) => `Name of ${id}`;

  it('should generate a pdf header and table', () => {
    const pollResult = {
      poll: mockPoll({
        content: {
          startTime: '2020-01-01T03:33:55Z',
          endTime: '2020-01-01T03:34:55Z',
        },
      }),
      results: {
        votes: {
          '@user-alice': '1',
        },
      },
      votingRights: ['@user-alice'],
    };
    const roomMember = mockRoomMember();

    expect(
      createPollPdfDefinition({
        authorName: 'Author Name',
        roomName: 'Room Name',
        pollResults: [pollResult],
        roomMemberEvents: [roomMember],
        getUserDisplayName,
      }),
    ).toMatchSnapshot();
  });
});
