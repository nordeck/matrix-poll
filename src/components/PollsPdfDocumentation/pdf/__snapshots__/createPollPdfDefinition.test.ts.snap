// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createPollPdfDefinition > should generate a pdf header and table 1`] = `
{
  "content": [
    [
      {
        "marginTop": 20,
        "pageBreak": undefined,
        "stack": [
          [
            {
              "color": "#444",
              "style": "tableExample",
              "table": {
                "body": [
                  [
                    {
                      "alignment": "left",
                      "fillColor": "#aaa",
                      "style": "tableHeader",
                      "text": "1 - My Title",
                    },
                    {
                      "alignment": "center",
                      "fillColor": "#aaa",
                      "style": "tableHeader",
                      "text": "Open poll",
                    },
                  ],
                  [
                    [
                      {
                        "alignment": "left",
                        "decoration": "underline",
                        "lineHeight": 1,
                        "margin": [
                          0,
                          2,
                        ],
                        "text": "General information",
                      },
                      {
                        "alignment": "left",
                        "lineHeight": 1.5,
                        "text": "Jan 1, 3:33 AM - 3:34 AM",
                      },
                      {
                        "alignment": "left",
                        "decoration": "underline",
                        "lineHeight": 1,
                        "margin": [
                          0,
                          2,
                        ],
                        "text": "Voting information",
                      },
                      {
                        "style": "list",
                        "ul": [
                          {
                            "alignment": "left",
                            "text": "Voting persons: 1",
                          },
                        ],
                      },
                    ],
                    [
                      [
                        {
                          "alignment": "left",
                          "decoration": "underline",
                          "margin": [
                            0,
                            2,
                          ],
                          "text": "Total voting result",
                        },
                        {
                          "alignment": "left",
                          "margin": [
                            0,
                            2,
                          ],
                          "text": "Cast votes: 1",
                        },
                        [
                          {
                            "style": "list",
                            "ul": [
                              {
                                "alignment": "left",
                                "text": "Yes: 1",
                              },
                            ],
                          },
                          {
                            "style": "list",
                            "ul": [
                              {
                                "alignment": "left",
                                "text": "No: 0",
                              },
                            ],
                          },
                          {
                            "style": "list",
                            "ul": [
                              {
                                "alignment": "left",
                                "text": "Invalid: 0",
                              },
                            ],
                          },
                        ],
                      ],
                    ],
                  ],
                  [
                    {
                      "border": [
                        true,
                        false,
                        true,
                        false,
                      ],
                      "colSpan": 2,
                      "decoration": "underline",
                      "text": "Description",
                    },
                  ],
                  [
                    {
                      "border": [
                        true,
                        false,
                        true,
                        true,
                      ],
                      "colSpan": 2,
                      "margin": [
                        0,
                        3,
                      ],
                      "text": "My Description",
                    },
                  ],
                  [
                    {
                      "border": [
                        true,
                        false,
                        true,
                        false,
                      ],
                      "colSpan": 2,
                      "decoration": "underline",
                      "text": "Voting question",
                    },
                  ],
                  [
                    {
                      "alignment": "left",
                      "bold": true,
                      "border": [
                        true,
                        false,
                        true,
                        true,
                      ],
                      "colSpan": 2,
                      "fontSize": 14,
                      "margin": [
                        0,
                        3,
                      ],
                      "style": "tableHeader",
                      "text": "My Question",
                    },
                  ],
                ],
                "headerRows": 1,
                "widths": [
                  "65%",
                  "35%",
                ],
              },
            },
          ],
          [],
          [],
        ],
      },
    ],
  ],
  "footer": [Function],
  "header": [Function],
  "info": {
    "author": "Author Name",
    "title": "Room Name",
  },
  "pageMargins": [
    40,
    80,
    40,
    40,
  ],
  "pageSize": "A4",
  "styles": {
    "list": {
      "margin": [
        5,
        0,
        0,
        0,
      ],
    },
    "tableBody": {
      "alignment": "center",
    },
    "tableHeader": {
      "alignment": "center",
      "bold": true,
      "fontSize": 13,
    },
  },
  "version": "1.5",
}
`;
